sympy/solvers/diophantine.py

```diff
@@ -2341,33 +2341,67 @@ def norm(u, w, a, b):
     return sqrt(dot((u_1, u_2), (u_1, u_2), w, a, b))
 
 
-def holzer(x_0, y_0, z_0, a, b, c):
+def holzer(x, y, z, a, b, c):
     """
-    Simplify the solution `(x_{0}, y_{0}, z_{0})` of the equation
-    `ax^2 + by^2 = cz^2` with `a, b, c > 0` and `z_{0}^2 \geq \mid ab \mid` to
-    a new reduced solution `(x, y, z)` such that `z^2 \leq \mid ab \mid`.
-    """
-    while z_0 > sqrt(a*b):
-
-        if c % 2 == 0:
-            k = c // 2
-            u_0, v_0 = base_solution_linear(k, y_0, -x_0)
+    Simplify the solution `(x, y, z)` of the equation
+    `ax^2 + by^2 = cz^2` with `a, b, c > 0` and `z^2 \geq \mid ab \mid` to
+    a new reduced solution `(x', y', z')` such that `z'^2 \leq \mid ab \mid`.
 
-        else:
-            k = 2*c
-            u_0, v_0 = base_solution_linear(c, y_0, -x_0)
+    The algorithm is an interpretation of Mordell's reduction as described
+    on page 8 of Cremona and Rusin's paper [1]_ and the work of Mordell in
+    reference [2]_.
 
-        w = -(a*u_0*x_0 + b*v_0*y_0) // (c*z_0)
+    References
+    ==========
 
-        if c % 2 == 1:
-            if w % 2 != (a*u_0 + b*v_0) % 2:
-                w = w + 1
+    .. [1] Efficient Solution of Rational Conices, J. E. Cremona and D. Rusin,
+           Mathematics of Computation, Volume 00, Number 0.
+    .. [2] Diophantine Equations, L. J. Mordell, page 48.
 
-        x = (x_0*(a*u_0**2 + b*v_0**2 + c*w**2) - 2*u_0*(a*u_0*x_0 + b*v_0*y_0 + c*w*z_0)) // k
-        y = (y_0*(a*u_0**2 + b*v_0**2 + c*w**2) - 2*v_0*(a*u_0*x_0 + b*v_0*y_0 + c*w*z_0)) // k
-        z = (z_0*(a*u_0**2 + b*v_0**2 + c*w**2) - 2*w*(a*u_0*x_0 + b*v_0*y_0 + c*w*z_0)) // k
+    """
 
+    if _odd(c):
+        k = 2*c
+    else:
+        k = c//2
+    k = Integer(k)
+
+    small = a*b*c
+    step = 0
+    while True:
+        t1, t2, t3 = a*x**2, b*y**2, c*z**2
+        # check that it's a solution
+        if t1 + t2 != t3:
+            if step == 0:
+                raise ValueError('bad starting solution')
+            break
         x_0, y_0, z_0 = x, y, z
+        if max(t1, t2, t3) <= small:
+            # Holzer condition
+            break
+
+        uv = u, v = base_solution_linear(k, y_0, -x_0)
+        if None in uv:
+            break
+
+        p, q = -(a*u*x_0 + b*v*y_0), c*z_0
+        r = Rational(p, q)
+        if _even(c):
+            w = _nint_or_floor(p, q)
+            assert abs(w - r) <= S.Half
+        else:
+            w = p//q  # floor
+            if _odd(a*u + b*v + c*w):
+                w += 1
+            assert abs(w - r) <= S.One
+
+        A = (a*u**2 + b*v**2 + c*w**2)
+        B = (a*u*x_0 + b*v*y_0 + c*w*z_0)
+        x = (x_0*A - 2*u*B) / k
+        y = (y_0*A - 2*v*B) / k
+        z = (z_0*A - 2*w*B) / k
+        assert all(i.is_Integer for i in (x, y, z))
+        step += 1
```

Comment:
Yes, there were some issues with the former implementation. I just reinterpreted the reference and what I've committed is that interpretation.  Here are some examples of failures in the former version:

```
>>> f = lambda eq: _diop_ternary_quadratic_normal((x, y, z), eq.as_coefficients_dict())
>>> f(92*x**2 - 99*y**2 - z**2)
(28557, 21296, 155058)  <-- should be set([(9, 7, 51)])
>>> f(234*x**2 - 65601*y**2 - z**2)
(5945, 31, 126819) <-- should be set([(233, 1, 3555)])
```